### DISCUSSION TITLE
Best practice for generics with Custom Models using pocketbase as a framework

### DISCUSSION THREAD

**Discussion by Trini8ed** [pocketbase/pocketbase#1889] [Q&A] [answered]
I've seen the examples on custom models and have those working, but I have not quite been able to figure out how to get custom models working with Generics. Any insights on or best practices would be appreciated if it's been done before.

The problem that generics solve for me is avoiding to re-write a bunch of functions that do the same thing. For example, let's say I have 20 different custom record model types and want to bind them to the defined data struct. I would have to write a query and find a method for each custom model.

Keep in mind I know the code below is not valid, but just trying to show what I would like to do and see if it's possible in-case others have any suggestions.

![image](https://user-images.githubusercontent.com/1361448/220198603-d6227efa-ca18-47e1-bee5-8a412dc3933d.png)

**Comment by ganigeorgiev** [comment, marked as answer]
I might misunderstood the screenshot, but I think the above should work if you change `any` to `models.Model` interface. For `Query` this could look something like:

```go
func Query[T models.Model](dao *daos.Dao) *dbx.SelectQuery {
	var m T
	return dao.ModelQuery(m)
}

// ...

Query[*database.Quantity](dao)
```

**Comment by Trini8ed** [comment]
Seems to be a bit closer to working but not quite there yet. I abstracted the functions into a single package here so it should all be readable from this snippet.

```go
var _ models.Model = (*Model)(nil)
var _ models.Model = (*Quantity)(nil)

func (m *Model) TableName() string {
	return "models"
}

func (m *Quantity) TableName() string {
	return "quantities"
}

type Model struct {
	models.BaseModel

	Name  string  `db:"name"`
	Data  string  `db:"data"`
	Image string  `db:"image"`
	Units string  `db:"units"` // "mm", "cm", "in",
	Scale float64 `db:"scale"` // 1.0, 10.0, 25.4
}

type Quantity struct {
	models.BaseModel
	ModelID string `db:"model_id"`

	Quantity int `db:"quantity"` // 1, 10, 1000, 10000
}

func Query[T models.Model](dao *daos.Dao) *dbx.SelectQuery {
	var m T
	return dao.ModelQuery(m)
}

func FindAllRecords[T models.Model](dao *daos.Dao) ([]*T, error) {
	records := []*T{}
	if err := Query[T](dao).
		All(records); err != nil {
		return nil, err
	}
	return records, nil
}

// ...

models, err := FindAllRecords[*Model](dao)
if err != nil {
	fmt.Println(err)
}
quantities, err := FindAllRecords[*Quantity](dao)
if err != nil {
	fmt.Println(err)
}

fmt.Println(models)
fmt.Println(quantities)
```

![image](https://user-images.githubusercontent.com/1361448/220207516-3b0d8d57-866e-407c-81d0-8b6afbe41747.png)

**Reply by ganigeorgiev** [reply]
When scanning the db rows into the custom struct (aka. `One()`, `All()`, etc.), the passed variable must be a pointer so that it can mutate it. Or in other words, you can try changing `FindAllRecords` to:

```go
func FindAllRecords[T models.Model](dao *daos.Dao) ([]T, error) {
	records := []T{}

	if err := Query[T](dao).All(&records); err != nil {
		return nil, err
	}

	return records, nil
}
```

**Reply by Trini8ed** [reply]
There we go excellent :)
